rule no-undefined-types does not work

eslint version: 5.9.0
eslint-plugin-jsdoc version: 3.9.1
Config:
```
'jsdoc/check-param-names': 1,
'jsdoc/check-tag-names': 1,
'jsdoc/check-types': 1,
'jsdoc/newline-after-description': 1,
'jsdoc/no-undefined-types': 1,
'jsdoc/require-description': 0,
'jsdoc/require-description-complete-sentence': 0,
'jsdoc/require-example': 0,
'jsdoc/require-hyphen-before-param-description': 1,
'jsdoc/require-param': 1,
'jsdoc/require-param-description': 1,
'jsdoc/require-param-name': 1,
'jsdoc/require-param-type': 1,
'jsdoc/require-returns-description': 1,
'jsdoc/require-returns-type': 1,
'jsdoc/valid-types': 1
```

Your own example:

```js
const MyType = require('my-library').MyType;

/**
 * @param {MyType} foo - Bar.
 */
function quux(foo) {

}
```
gives the error `warning  The type 'MyType' is undefined  jsdoc/no-undefined-types`

@brettz9 unfortunately the problem remains.

```bash
$ eslint/bin/eslint.js -v
v5.16.0
```

Run check with the [config.js](https://pastebin.com/xDMDmTpZ):

```bash
$ eslint/bin/eslint.js -c config.js test.js
v5.16.0
```

Result:

```
test.js
  3:7   error    'MyType' is assigned a value but never used  no-unused-vars
  6:0   warning  The type 'MyType' is undefined               jsdoc/no-undefined-types
  8:10  error    'quux' is defined but never used             no-unused-vars
  8:17  error    'foo' is defined but never used              no-unused-vars
  8:23  error    Unexpected empty function 'quux'             no-empty-function

✖ 5 problems (4 errors, 1 warning)
```

There are some other unrelated expected issues here, please ignore them.

--- a/README.md
+++ b/README.md
@@ -1391,6 +1391,15 @@ const MyType = require('my-library').MyType;
 
 }
 
+const MyType = require('my-library').MyType;
+
+/**
+ * @param {MyType} foo - Bar.
+ */
+  function quux(foo) {
+
+}
+
 import {MyType} from 'my-library';
 
 /**

--- a/README.md
+++ b/README.md
@@ -1371,6 +1371,13 @@ function quux(foo) {
 
 }
 
+/**
+ * @param {Promise} foo - Bar.
+ */
+function quux(foo) {
+
+}
+
 class MyClass {}
 
 /**
@@ -1411,7 +1418,7 @@ import {MyType} from 'my-library';
 
 }
 
-/*global MyType*/
+/*globals MyType*/
 
 /**
  * @param {MyType} foo - Bar.

--- a/src/rules/noUndefinedTypes.js
+++ b/src/rules/noUndefinedTypes.js
@@ -51,9 +51,20 @@ export default iterateJsdoc(({
   })
 
     // If the file is a module, concat the variables from the module scope.
-    .concat(scopeManager.isModule() ? globalScope.childScopes[0].variables.map((variable) => {
-      return variable.name;
-    }) : [])
+    .concat(
+
+      // This covers `commonjs` as well as `node`
+      scopeManager.__options.nodejsScope ||
+      scopeManager.isModule() ?
+        globalScope.childScopes.reduce((arr, {variables}) => {
+          // Flatten
+          arr.push(...variables);
+
+          return arr;
+        }, []).map(({name}) => {
+          return name;
+        }) : []
+    )
     .concat(extraTypes)
     .concat(typedefDeclarations);
 

--- a/test/rules/assertions/noUndefinedTypes.js
+++ b/test/rules/assertions/noUndefinedTypes.js
@@ -58,7 +58,25 @@ export default {
           function quux(foo) {
 
         }
-      `
+      `,
+      env: {
+        node: true
+      }
+    },
+    {
+      code: `
+        const MyType = require('my-library').MyType;
+
+        /**
+         * @param {MyType} foo - Bar.
+         */
+          function quux(foo) {
+
+        }
+      `,
+      env: {
+        node: false
+      }
     },
     {
       code: `

--- a/test/rules/assertions/noUndefinedTypes.js
+++ b/test/rules/assertions/noUndefinedTypes.js
@@ -33,6 +33,19 @@ export default {
     },
     {
       code: `
+          /**
+           * @param {Promise} foo - Bar.
+           */
+          function quux(foo) {
+
+          }
+      `,
+      env: {
+        es6: true
+      }
+    },
+    {
+      code: `
           class MyClass {}
 
           /**
@@ -97,7 +110,7 @@ export default {
     },
     {
       code: `
-        /*global MyType*/
+        /*globals MyType*/
 
         /**
          * @param {MyType} foo - Bar.

